src/binding.cc

```diff
@@ -0,0 +1,162 @@
+    throw std::runtime_error("result length exceeds uint32 max");
+  }
+  for (uint32_t i = 0; i < vec.size(); i++) {
+    ret[i] = Buffer<BYTE>::Copy(env, vec[i].data(), vec[i].size());
+  }
+  return ret;
+}
+
+// Convert UTF-8 to a Windows UTF-16 wstring.
+std::wstring MultiByteToWideChar(Value value) {
+  static_assert(sizeof(std::wstring::value_type) == sizeof(std::u16string::value_type),
+      "wstring and u16string have the same value type on Windows");
+  std::u16string u16 = value.ToString();
+  return std::wstring(u16.begin(), u16.end());
+}
+
+Value ExportAllCertificatesSync(const CallbackInfo& args) {
+  std::wstring sys_store_name = MultiByteToWideChar(args[0].ToString());
+  DWORD store_type = args[1].ToNumber().Uint32Value();
+
+  try {
+    return BufferListToArray(
+        args.Env(), ExportAllCertificates(sys_store_name, store_type));
+  } catch (const std::exception& e) {
+    throw Error::New(args.Env(), e.what());
+  }
+}
+
+ExportCertificateAndKeyArgs GatherExportCertificateAndKeyArgs(const CallbackInfo& args) {
+  ExportCertificateAndKeyArgs exp_args;
+  exp_args.password_buf = MultiByteToWideChar(args[0].ToString());
+  exp_args.sys_store_name = MultiByteToWideChar(args[1].ToString());
+  exp_args.store_type = args[2].ToNumber().Uint32Value();
+  exp_args.require_private_key = args[4].ToBoolean();
+
+  Object search_spec = args[3].ToObject();
+  if (search_spec.HasOwnProperty("thumbprint")) {
+    exp_args.use_thumbprint = true;
+    Buffer<BYTE> thumbprint = search_spec.Get("thumbprint").As<Buffer<BYTE>>();
+    exp_args.thumbprint = {thumbprint.Data(), thumbprint.Data() + thumbprint.Length()};
+  } else if (search_spec.HasOwnProperty("subject")) {
+    exp_args.use_thumbprint = false;
+    exp_args.subject = MultiByteToWideChar(search_spec.Get("subject").ToString());
+  } else {
+    throw Error::New(args.Env(), "Need to specify either `thumbprint` or `subject`");
+  }
+  return exp_args;
+}
+
+// Export a given certificate from a system certificate store,
+// identified either by its thumbprint or its subject line.
+Value ExportCertificateAndKeySync(const CallbackInfo& args) {
+  ExportCertificateAndKeyArgs exp_args = GatherExportCertificateAndKeyArgs(args);
+  try {
+    auto result = ExportCertificateAndKey(exp_args);
+    return Buffer<BYTE>::Copy(args.Env(), result.data(), result.size());
+  } catch (const std::exception& e) {
+    throw Error::New(args.Env(), e.what());
+  }
+}
+
+Value ExportAllCertificatesAsync(const CallbackInfo& args) {  
+  class Worker final : public AsyncWorker {
+    public:
+      Worker(Function callback, std::wstring&& sys_store_name, DWORD store_type)
+        : AsyncWorker(callback, PACKAGE ":ExportAllCertificates"),
+          sys_store_name(std::move(sys_store_name)), store_type(store_type) {}
+      ~Worker() {}
+
+      void Execute() override {
+        results = ExportAllCertificates(sys_store_name, store_type);
+      }
+
+      void OnOK() override {
```

Comment:
I was looking at the documentation for [AsyncWorker](https://github.com/nodejs/node-addon-api/blob/main/doc/async_worker.md) and there is an OnError method that we might want to override here I think. Probably we want to have some logging or at least a friendly error message in case of whatever error might happen during the export of all the certificates, what do you think?

Reply:
The default implementation of `OnError` calls the callback with a JS Error argument built from the C++ exception that was thrown, if there was one. That should work fine, right?

Reply:
Yeah I think that is perfect. Just wanted to make sure it wasn't shallowing the error or panicking.